# 每天一个小知识点
## 第十八天 
---

## 1、攻击类型
XSS攻击: 跨站脚本攻击，是一种代码注入攻击。攻击者把可执行的恶意脚本注入搭配页面中,使之在用户的浏览器上运行,从而盗取用户的信息如 cookie 等。  
解决方案：   
1、对需要插入到 HTML 中的代码做好充分的转义  
2、使用 CSP，建立一个白名单，告诉浏览器哪些外部资源可以加载和执行,从而防止恶意代码的注入攻击 

CSRF攻击: 跨站请求伪造攻击,攻击者诱导用户进入一个第三方网站,然后该网站向被攻击网站发送跨站请求,如果用户在被攻击网站中保存了登录状态,那么攻击者就可以利用这个登录状态,绕过后台的用户验证,冒充用户向服务器执行一些操作。比如攻击者可以通过在输入留言框内输入可发送请求的代码，留言放到网站上，这样每个人进入这个网站的时候都会执行这行代码并发送携带自己的cookie账号密码等关键信息的请求，这样攻击者就能拿到用户的信息伪造拿去登录了  
解决方案：  
1.同源检测  
2.使用 CSRF Token 进行验证  
3.限制 cookie 不能作为被第三方使用

## 2、实现一个countFn,接受一个函数作为参数,返回一个新函数,具备调用计数器功能
1、proxy  
proxy中的函数调用操作的捕捉器是apply,因此我们需要有一个计数的属性count,然后重写apply中的内容,在这做count的增量就可以了,这样每次调用testCount(),都会执行apply里的内容.至于count,不论放在handler这个对象还是放在countFn这个函数里都可以
```js
function countFn(fn){
  var count = 0
  let handler = {
    // count: 0
    apply: function(target, that, agrs) {
      console.log(++count)
      target.apply(that, args)
    }
  }
  return new Proxy(fn, handler)
}
function test () {
  console.log('test')
  console.log(this)
}
function test2 () {
  console.log('test2')
}
const testCount = countFn(test)
const test2Count = countFn(test2)
testCount() // 1 'test'
testCount() // 2 'test'
testCount() // 3 'test'

test2Count() // 1 'test2'
test2Count() // 2 'test2'
test2Count() // 3 'test2'
```
2、闭包  
```js
function countFn(fn) {
  var count = 0
  return function() {
    count++
    fn.call(this,...arguments)
    console.log(count)
  }
}

function fn1(test) {
  console.log('test1', test)
}
function fn2(test) {
  console.log('test2', test)
}

var test1 = countFn(fn1)
var test2 = countFn(fn2)
test1('111')//test1 111 1
test2('111')//test2 111 1
test1('222')//test1 222 2
test2('222')//test2 222 2
test2('333')//test2 333 3
test1('333')//test1 333 3

```
3、放在原型链上
```js
Function.prototype.countFn=function(){
    var count=0;
    var fn=this
    return function(){
        count++
        fn.call(this,...arguments)
        console.log(count)
    }
}
var test1= fn1.countFn()
var test2= fn2.countFn()
test1('111')//test1 111 1
test2('111')//test2 111 1
test1('222')//test1 222 2
test2('222')//test2 222 2
test2('333')//test2 333 3
test1('333')//test1 333 3
```

## 3、Object.creat(null)和直接使用字面量{}创建空对象有什么区别和好处
Object.create()有两个参数,第一个参数是指定的原型对象,第二个参数是可选参数,给新对象自身添加新属性及描述器  
直接创建{},创建好的空对象会有很多对象内置的原型属性和方法,而利用Object.create(null)创建的空对象,相比较没有对象原型上的属性和方法.这样可避免新增的方法污染原型上的方法和属性,此外,还可以节省hasOwnProperty带来的性能损失
## 4、手写Promise.then如何保证后一个then里的方法在前一个then结束之后再执行？
我们可以将传给`then`的函数和新`promise`的`resolve`一起`push`到前一个`promise`的`callbacks`数组中,达到承前启后的效果:  
承前: 当前一个`promise`完成后,调用其`resolve`变更状态,在这个`resolve`里会依次调用`callbacks`里的回调,这样就执行了`then`里的方法了  
启后: 上一步中,当`then`里的方法执行完成后,返回一个结果,如果这个结果是个简单的值,就直接调用新的`promise`的`resolve`,让其变更状态,这又会调用新`promise`的`callbacks`数组里的方法,循环往复....如果返回的结果是个`promise`,则需要等它完成之后再触发新`promise`的`resolve`,所以可以在其结果的`then`里调用新`promise`的`resolve`
```js
then(onFulfilled, onReject) {
  //保留前一个promise的this
  const self = this
  return new myPromise((resolve, reject) => {
    //封装前一个promise成功时执行的函数
    let fulfilled = () => {
      try {
        const result = onFulfilled(self.value) //承前
        return resulte instanceof myPromise?result.then(resolve,reject):resolve(result) //启后
      } catch(err){
        reject(err)
      }
    }
    // 封装前一个promise失败时执行的函数
    let rejected = () => {
      try{
        const result = onReject(self.reason)
        return result instanceof MyPromise? result.then(resolve, reject) :reject(result)
        } catch(err){
          reject(err)
        }
      }
    }
    switch(self.status){
      case PENDING: 
        self.onFulfilledCallbacks.push(fulfilled);
        self.onRejectedCallbacks.push(rejected);
        break;
      case FULFILLED:
        fulfilled();
        break;
      case REJECT:
        rejected();
        break;
    }
  })
}
```
注意:  
连续多个 then 里的回调方法是同步注册的，但注册到了不同的 callbacks 数组中，因为每次 then 都返回新的 promise 实例（参考上面的例子和图）  
注册完成后开始执行构造函数中的异步事件，异步完成之后依次调用 callbacks 数组中提前注册的回调
